Merge Geth v1.13.15

AnnaShaleva opened the pull request:
This PR merges [v1.13.15](https://github.com/ethereum/go-ethereum/releases/tag/v1.13.15) of Geth into bane-main and aimed to close #188.

Close #291.

- [X]  The node functions normally and consensus runs fine (1/4/7 privnet).
- [X] Mainnet/T4 sync is OK.

UPD on the compatibility problem described earlier: there's actually no compatibility problems with the 1.13.15 version merge. Turns out that Geth v1.13.15 was an unclean release from non-master branch. It's based on v1.13.14 with five additional commits cherry-picked (ref. https://github.com/ethereum/go-ethereum/pull/29556, the release branch itself is https://github.com/ethereum/go-ethereum/tree/release/1.13).

The list of important PRs that were merged. @txhsl, @chenquanyu, please take a look at these PRs before the review.
* Fix the enforced minimum miner tip (https://github.com/ethereum/go-ethereum/pull/28933).
* Add Cancun timestamps for mainnet (https://github.com/ethereum/go-ethereum/pull/28958). This PR may be useful for us while configuring our own schedule for Cancun.
* Genesis allocations were moved to a separate package (https://github.com/ethereum/go-ethereum/pull/29003). This PR does not affect us directly, but we need to be aware of this change.
* **An ability to pre-filter mempooled transactions by dynamic fee is added** (https://github.com/ethereum/go-ethereum/pull/29005). This change affects the #166, it needs to be adjusted accordingly.
* **Override block's BaseFee in eth_call API** (https://github.com/ethereum/go-ethereum/pull/29051). This change seems not to affect us directly, but it would be nice if you double-check it.


Also, it would be nice if you double-check the merged commits, maybe you'll find something important that we should react to.

karalabe committed:
params: release Geth v1.15.5

fjl committed:
Merge branch 'release/1.13' into release-1.13.6

fjl committed:
all: release go-ethereum v1.13.6

fjl committed:
all: release go-ethereum v1.13.7

holiman committed:
all: release go-ethereum v1.13.8

Release 1.13.8

fjl committed:
Merge branch 'master' into release/1.13

karalabe committed:
Merge branch 'master' into release/1.13

holiman committed:
all: release go-ethereum v1.13.11 (#28868)

Release 1.13.11

holiman committed:
params: begin v.1.13.12 release cycle

bodhi-crypo committed:
internal/flags: fix typo (#28876)

protolambda committed:
core/types: fix and test handling of faulty nil-returning signer (#28879)

This adds an error if the signer returns a nil value for one of the signature value fields.

keienWang committed:
README.md: fix travis badge (#28889)

The hyperlink in the README file that directs to the Travis CI build was broken.
This commit updates the link to point to the corrent build page.

lightclient committed:
eth/catalyst: allow payload attributes v1 in fcu v2 (#28882)

At some point, `ForkchoiceUpdatedV2` stopped working for `PayloadAttributesV1` while `paris` was active. This was causing a few failures in hive. This PR fixes that, and also adds a gate in `ForkchoiceUpdatedV1` to disallow `PayloadAttributesV3`.

keienWang committed:
docs/postmortems: fix outdated link (#28893)

rjl493456442 committed:
core: reset tx lookup cache if necessary (#28865)

This pull request resets the txlookup cache if chain reorg happens, 
preventing them from remaining reachable. It addresses failures in
the hive tests.

holiman committed:
build: fix problem with windows line-endings in CI download (#28900)

fixes #28890

rjl493456442 committed:
eth/downloader: fix skeleton cleanup (#28581)

* eth/downloader: fix skeleton cleanup

* eth/downloader: short circuit if nothing to delete

* eth/downloader: polish the logic in cleanup

* eth/downloader: address comments

holiman committed:
deps: update memsize (#28916)

karalabe committed:
core/txpool/blobpool: post-crash cleanup and addition/removal metrics (#28914)

* core/txpool/blobpool: clean up resurrected junk after a crash

* core/txpool/blobpool: track transaction insertions and rejections

* core/txpool/blobpool: linnnnnnnt

karalabe committed:
core/txpool: don't inject lazy resolved transactions into the container (#28917)

* core/txpool: don't inject lazy resolved transactions into the container

* core/txpool: minor typo fixes

zoereco committed:
core/types: fix typo (#28922)

ziogaschr committed:
p2p: fix accidental termination of portMappingLoop (#28911)

karalabe committed:
internal/flags: fix --miner.gasprice default listing (#28932)

rex4539 committed:
all: fix typos in comments (#28881)

Halimao committed:
Makefile: add help target to display available targets (#28845)


Co-authored-by: Martin HS <martin@swende.se>
Co-authored-by: Felix Lange <fjl@twurst.com>

rjl493456442 committed:
core: cache transaction indexing tail in memory (#28908)

karalabe committed:
eth, miner: fix enforcing the minimum miner tip (#28933)

* eth, miner: fix enforcing the minimum miner tip

* ethclient/simulated: fix failing test due the min tip change

* accounts/abi/bind: fix simulater gas tip issue

lmittmann committed:
core/state, core/vm: minor uint256 related perf improvements (#28944)

lightclient committed:
cmd,internal/era: implement `export-history` subcommand (#26621)

* all: implement era format, add history importer/export

* internal/era/e2store: refactor e2store to provide ReadAt interface

* internal/era/e2store: export HeaderSize

* internal/era: refactor era to use ReadAt interface

* internal/era: elevate anonymous func to named

* cmd/utils: don't store entire era file in-memory during import / export

* internal/era: better abstraction between era and e2store

* cmd/era: properly close era files

* cmd/era: don't let defers stack

* cmd/geth: add description for import-history

* cmd/utils: better bytes buffer

* internal/era: error if accumulator has more records than max allowed

* internal/era: better doc comment

* internal/era/e2store: rm superfluous reader, rm superfluous testcases, add fuzzer

* internal/era: avoid some repetition

* internal/era: simplify clauses

* internal/era: unexport things

* internal/era,cmd/utils,cmd/era: change to iterator interface for reading era entries

* cmd/utils: better defer handling in history test

* internal/era,cmd: add number method to era iterator to get the current block number

* internal/era/e2store: avoid double allocation during write

* internal/era,cmd/utils: fix lint issues

* internal/era: add ReaderAt func so entry value can be read lazily

Co-authored-by: lightclient <lightclient@protonmail.com>
Co-authored-by: Martin Holst Swende <martin@swende.se>

* internal/era: improve iterator interface

* internal/era: fix rlp decode of header and correctly read total difficulty

* cmd/era: fix rebase errors

* cmd/era: clearer comments

* cmd,internal: fix comment typos

---------

Co-authored-by: Martin Holst Swende <martin@swende.se>

lightclient committed:
core,params: add holesky to default genesis function (#28903)

fjl committed:
node, rpc: add configurable HTTP request limit (#28948)

Adds a configurable HTTP request limit, and bumps the engine default

zoereco committed:
all: fix docstring names (#28923)

* fix wrong comment

* reviewers input

* Update log/handler_glog.go

---------

Co-authored-by: Martin HS <martin@swende.se>

bodhi-crypo committed:
ethclient/simulated: fix typo (#28952)

(ethclient/simulated):fix typo

fjl committed:
eth/gasprice: fix percentile validation in eth_feeHistory (#28954)

karalabe committed:
cmd/devp2p, eth: drop support for eth/67 (#28956)

lightclient committed:
params, core/forkid: add mainnet timestamp for Cancun (#28958)

* params: add cancun timestamp for mainnet

* core/forkid: add test for mainnet cancun forkid

* core/forkid: update todo tests for cancun

s1na committed:
internal/ethapi: add support for blobs in eth_fillTransaction (#28839)

This change adds support for blob-transaction in certain API-endpoints, e.g. eth_fillTransaction. A follow-up PR will add support for signing such transactions.

lightclient committed:
internal/era: update block index format to be based on record offset (#28959)

As mentioned in #26621, the block index format for era1 is not in line with the regular era block index. This change modifies the index so all relative offsets are based against the beginning of the block index record.

holiman committed:
params: go-ethereum v1.13.12 stable

holiman committed:
all: release go-ethereum v1.13.12 (#28961)

all: release go-ethereum v1.13.12

holiman committed:
params: begin v1.13.13 release cycle

holiman committed:
build: remove ubuntu 'lunar' build (#28962)

krauspt committed:
fix: update outdated link to trezor docs (#28966)

fix: update link to trezor

mask-pp committed:
internal/ethapi: fix gas estimation bug in eth_fillTransaction for blob tx (#28929)

weiihann committed:
core/txpool/legacypool: use uint256.Int instead of big.Int (#28606)

This change makes the legacy transaction pool use of `uint256.Int` instead of `big.Int`. The changes are made primarily only on the internal functions of legacypool. 

---------

Co-authored-by: Martin Holst Swende <martin@swende.se>

lindlof committed:
internal/ethapi, signer/core: fix documentation-links (#28979)

fix: management api links

rjl493456442 committed:
all: remove the dependency from trie to triedb (#28824)

This change removes the dependency from trie package to triedb package.

holiman committed:
cmd/utils: fix merge-breakage in test (#28985)

holiman committed:
tests: fix goroutine leak related to state snapshot generation (#28974)


---------

Co-authored-by: Felix Lange <fjl@twurst.com>

mask-pp committed:
ethereum, ethclient: add blob transaction fields in CallMsg (#28989)


Co-authored-by: Felix Lange <fjl@twurst.com>

bkellerman committed:
core/txpool/blobpool: rename variables in comments (#28981)


Co-authored-by: Felix Lange <fjl@twurst.com>

colinlyguo committed:
cmd/devp2p: fix modulo in makeBlobTxs (#28970)

lightclient committed:
eth/catalyst,beacon/engine: implement GetClientVersionV1 (#28915)

holiman committed:
tests: update execution spec tests + split statetest exec (#28993)

MariusVanDerWijden committed:
eth/catalyst: add getClientVersion to capabilities (#28994)

bodhi-crypo committed:
cmd/evm: fix typo in test script (#28995)

fjl committed:
cmd/devp2p/internal/ethtest: some fixes for the eth test suite (#28996)

Improving two things here:

On hive, where we look at these tests, the Go code comment above the test
is not visible. When there is a failure, it's not obvious what the test is actually
expecting. I have converted the comments in to printed log messages to
explain the test more.

Second, I noticed that besu is failing some tests because it happens to request
a header when we want it to send transactions. Trying the minimal fix here to
serve the headers.

Co-authored-by: lightclient <14004106+lightclient@users.noreply.github.com>

colinlyguo committed:
core/txpool/legacypool: remove a redundant heap.Init (#28910)


Co-authored-by: Martin HS <martin@swende.se>
Co-authored-by: Felix Lange <fjl@twurst.com>

karalabe committed:
core/txpool/blobpool: update the blob db with corruption handling (#29001)

Updates billy to a more recent version which is more robust in the face of corrupt data (e.g. after a hard crash)

s1na committed:
core: move genesis alloc types to core/types (#29003)

We want to use these types in public user-facing APIs, so they shouldn't be in core.

Co-authored-by: Felix Lange <fjl@twurst.com>

karalabe committed:
core/txpool, eth, miner: pre-filter dynamic fees during pending tx retrieval (#29005)

* core/txpool, eth, miner: pre-filter dynamic fees during pending tx retrieval

* miner: fix typo

* core/txpool: handle init-error in blobpool without panicing

---------

Co-authored-by: Martin Holst Swende <martin@swende.se>

rjl493456442 committed:
ethstats: prevent panic if head block is not available (#29020)

This pull request fixes a flaw in ethstats which can lead to node crash

A panic could happens when the local blockchain is reorging which causes the original head block not to be  reachable (since number->hash canonical mapping is deleted). In order to prevent the panic, the block nilness is now checked in ethstats.

cuiweixie committed:
core: using math.MaxUint64 instead of 0xffffffffffffffff (#29022)

karalabe committed:
core/txpool, miner: speed up blob pool pending retrievals (#29008)

* core/txpool, miner: speed up blob pool pending retrievals

* miner: fix test merge issue

* eth: same same

* core/txpool/blobpool: speed up blobtx creation in benchmark a bit

* core/txpool/blobpool: fix linter

---------

Co-authored-by: Martin Holst Swende <martin@swende.se>

holiman committed:
core/vm, params: ensure order of forks, prevent overflow (#29023)

This PR fixes an overflow which can could happen if inconsistent blockchain rules were configured. Additionally, it tries to prevent such inconsistencies from occurring by making sure that merge cannot be enabled unless previous fork(s) are also enabled.

karalabe committed:
core/txpool, eth, miner: retrieve plain and blob txs separately (#29026)

* core/txpool, eth, miner: retrieve plain and blob txs separately

* core/txpool: fix typo, no farming

* miner: farm all the typos

Co-authored-by: Martin HS <martin@swende.se>

---------

Co-authored-by: Martin HS <martin@swende.se>

buddh0 committed:
core/txpool: fix typo (#29031)

buddh0 committed:
core,eth,internal: fix typo (#29024)

tmelhao committed:
params: add cancun upgrade banner (#29042)

params: add cancun banner

Signed-off-by: tmelhao <tmelhao@gmail.com>
Co-authored-by: tmelhao <tmelhao@gmail.com>

colinlyguo committed:
core/txpool: fix typo (#29036)

* fix typos

* address comments

s1na committed:
internal/ethapi: fix defaults for blob fields (#29037)


Co-authored-by: Martin HS <martin@swende.se>

fjl committed:
params: release go-ethereum v1.13.13 stable

fjl committed:
Merge branch 'master' into release/1.13

fjl committed:
params: begin v1.13.14 release cycle

ArtificialPB committed:
internal/ethapi: use overriden baseFee for gasPrice (#29051)

eth_call and debug_traceCall allow users to override various block fields, among them base fee. However the overriden base fee was not considered for computing the effective gas price of that message, and instead base fee of the base block was used. This has been fixed in this commit.

colinlyguo committed:
internal/ethapi:fix zero rpc gas cap in eth_createAccessList (#28846)

This PR enhances eth_createAccessList RPC call to support scenarios where the node is launched with an unlimited gas cap (--rpc.gascap 0). The eth_createAccessList RPC call returns failure if user doesn't explicitly set a gas limit.

tmelhao committed:
eth/catalyst: fix wrong error message of payloadV2 after cancun (#29049)

* eth/catalyst: the same error format

Signed-off-by: tmelhao <tmelhao@gmail.com>

* eth/catalyst: wrong error message for payloadV2 post-cancun

Signed-off-by: tmelhao <tmelhao@gmail.com>

* eth/catalyst: parentBeaconBlockRoot -> parentBlockBeaconRoot

Signed-off-by: tmelhao <tmelhao@gmail.com>

* apply commit review

Signed-off-by: tmelhao <tmelhao@gmail.com>

---------

Signed-off-by: tmelhao <tmelhao@gmail.com>
Co-authored-by: tmelhao <tmelhao@gmail.com>

Commit:
core/txpool: reject blob txs with blob fee cap below the minimum (#29081)

* make blobpool reject blob transactions with fee below the minimum

* core/txpool: some minot nitpick polishes and unified error formats

* core/txpool: do less big.Int constructions with the min blob cap

---------

Co-authored-by: Péter Szilágyi <peterke@gmail.com>

0x2d3c committed:
p2p, log, rpc: use errors.New to replace fmt.Errorf with no parameters (#29074)

cuiweixie committed:
eth/catalyst: enable some commented-out testcases   (#29073)

mask-pp committed:
internal/ethapi: pass blob hashes to gas estimation (#29085)

justindhillon committed:
cmd/clef: add spaces in README.md table (#29077)

Add space after links in so they are clickable in vscode.

cuiweixie committed:
eth/catalyst: remove variable in tx conversion loop (#29076)

karalabe committed:
core/txpool/blobpool: reduce default database cap for rollout (#29090)

xcore/txpool/blobpool: reduce default database cap for rollout

jsvisa committed:
core/txpool: no need to log loud rotate if no local txs (#29083)

* core/txpool: no need to run rotate if no local txs

Signed-off-by: jsvisa <delweng@gmail.com>

* Revert "core/txpool: no need to run rotate if no local txs"

This reverts commit 17fab173883168c586d57ca9c05dfcbd9e7831b4.

Signed-off-by: jsvisa <delweng@gmail.com>

* use Debug if todo is empty

Signed-off-by: jsvisa <delweng@gmail.com>

---------

Signed-off-by: jsvisa <delweng@gmail.com>

andreisilviudragnea committed:
eth/tracers: Fix callTracer logs on onlyTopCall == true (#29068)

Commit:
core/txpool: elevate the 'already reserved' error into a constant (#29095)

declare the 'already reserved' error in errors.go

karalabe committed:
params: release Geth v1.13.14

karalabe committed:
Merge branch 'master' into release/1.13

rjl493456442 committed:
eth/protocols/snap: skip retrieval for completed storages (#29378)

* eth/protocols/snap: skip retrieval for completed storages

* eth/protocols/snap: address comments from peter

* eth/protocols/snap: add comments

holiman committed:
core/rawdb: add sanity-limit to header accessor (#29534)

holiman committed:
eth/filters: enforce topic-limit early on filter criterias (#29535)

This PR adds a limit of 1000 to the "inner" topics in a filter-criteria

rjl493456442 committed:
core, eth/protocols/snap, trie: fix cause for snap-sync corruption, implement gentrie (#29313)

This pull request defines a gentrie for snap sync purpose.

The stackTrie is used to generate the merkle tree nodes upon receiving a state batch. Several additional options have been added into stackTrie to handle incomplete states (either missing states before or after).

In this pull request, these options have been relocated from stackTrie to genTrie, which serves as a wrapper for stackTrie specifically for snap sync purposes.

Further, the logic for managing incomplete state has been enhanced in this change. Originally, there are two cases handled:

-    boundary node filtering
-    internal (covered by extension node) node clearing

This changes adds one more:

- Clearing leftover nodes on the boundaries.

This feature is necessary if there are leftover trie nodes in database, otherwise node inconsistency may break the state healing.

holiman committed:
params: release Geth v 1.13.15

AnnaShaleva committed:
Merge remote-tracking branch 'ethereum/release/1.13' into upd-v1.13.15

AnnaShaleva committed:
core: fix filing test after v1.13.15 merge

msg.GasTipCap is nil, but I'm not sure why, it shouldn't be like that
in the real network.

```
--- FAIL: TestExecutionSpecState (0.01s)
    --- FAIL: TestExecutionSpecState/berlin/eip2930_access_list/acl/access_list.json (0.00s)
        --- FAIL: TestExecutionSpecState/berlin/eip2930_access_list/acl/access_list.json/tests/berlin/eip2930_access_list/test_acl.py::test_access_list[fork_Berlin-state_test] (0.00s)
            --- FAIL: TestExecutionSpecState/berlin/eip2930_access_list/acl/access_list.json/tests/berlin/eip2930_access_list/test_acl.py::test_access_list[fork_Berlin-state_test]/Berlin/0/hash/trie (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5e2ad8]

goroutine 78973 [running]:
testing.tRunner.func1.2({0xd79340, 0x162a1b0})
	/usr/local/go/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1634 +0x377
panic({0xd79340?, 0x162a1b0?})
	/usr/local/go/src/runtime/panic.go:770 +0x132
math/big.(*Int).Cmp(0x0?, 0x0?)
	/usr/local/go/src/math/big/int.go:381 +0x18
github.com/ethereum/go-ethereum/core.(*StateTransition).preCheck(0xc001317628)
	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/core/state_transition.go:300 +0x485
github.com/ethereum/go-ethereum/core.(*StateTransition).TransitionDb(0xc001317628)
	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/core/state_transition.go:393 +0x51
github.com/ethereum/go-ethereum/core.ApplyMessage(0x0?, 0xf97800?, 0xf96f28?)
	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/core/state_transition.go:184 +0x57
github.com/ethereum/go-ethereum/tests.(*StateTest).RunNoVerify(0xc0001794a0, {{0xc000353730?, 0xc001142006?}, 0x7dc007fd8a68?}, {{0x0, 0x0}, 0x0, 0x0, {0x0, 0x0, ...}}, ...)
	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/tests/state_test_util.go:301 +0x9e5
github.com/ethereum/go-ethereum/tests.(*StateTest).Run(0xc0001794a0, {{0xc000353730?, 0x6?}, 0x0?}, {{0x0, 0x0}, 0x0, 0x0, {0x0, 0x0, ...}}, ...)
	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/tests/state_test_util.go:198 +0xf7
github.com/ethereum/go-ethereum/tests.execStateTest.func1.1({{0x0, 0x0}, 0x0, 0x0, {0x0, 0x0, 0x0}})
	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/tests/state_test.go:113 +0xef
github.com/ethereum/go-ethereum/tests.withTrace(0xc001129520, 0x4ef00, 0xc001317f18)
	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/tests/state_test.go:167 +0x75
github.com/ethereum/go-ethereum/tests.execStateTest.func1(0xc001129520)
	/home/anna/Documents/GitProjects/bane-labs/go-ethereum/tests/state_test.go:111 +0x12b
testing.tRunner(0xc001129520, 0xc002c0d350)
	/usr/local/go/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 78972
	/usr/local/go/src/testing/testing.go:1742 +0x390

```

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
eth: remove stale test

Support for eth/67 was dropped in 8a76a814.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
Merge pull request #234 from bane-labs/fix-tests-1.13.15

core: fix filing test after v1.13.15 merge

AnnaShaleva committed:
Merge branch 'bane-main' into upd-v1.13.15

AnnaShaleva commented:
Tests pass, but privnet isn't running:
```
INFO [08-09|19:38:45.069] HTTP server started                      endpoint=[::]:8562 auth=false prefix= cors= vhosts=*
INFO [08-09|19:38:45.070] WebSocket enabled                        url=ws://[::]:8572
INFO [08-09|19:38:45.071] WebSocket enabled                        url=ws://127.0.0.1:8552
INFO [08-09|19:38:45.071] HTTP server started                      endpoint=127.0.0.1:8552 auth=true  prefix= cors=localhost vhosts=localhost
INFO [08-09|19:38:46.638] Unlocked account                         address=0x74f4EFFb0B538BAec703346b03B6d9292f53A4CD
INFO [08-09|19:38:46.638] Legacy pool tip threshold updated        tip=0
INFO [08-09|19:38:46.638] Legacy pool tip threshold updated        tip=1,000,000,000
INFO [08-09|19:38:46.638] Initializing BFT consensus               account=0x74f4EFFb0B538BAec703346b03B6d9292f53A4CD
INFO [08-09|19:38:46.638] Snap sync complete, auto disabling
INFO [08-09|19:38:46.638] Fetching latest sealing proposal         "desired number"=1
INFO [08-09|19:38:46.639] Commit new sealing work                  number=1 sealhash=e66155..adaa87 "parent hash"=edba97..2eac00 etherbase=0x1212000000000000000000000000000000000003 txs=0 gas=0 fees=0 elapsed="171.163µs"
INFO [08-09|19:38:47.640] Sealing proposal updated                 number=1 sealhash=e66155..adaa87 "parent hash"=edba97..2eac00 txs=0
INFO [08-09|19:38:47.640] Starting dBFT engine                     "last height"=0 "last timestamp"=0
panic: failed to retrieve next block validators: failed to perform 'getCurrentConsensus' call: execution reverted

goroutine 4252 [running]:
github.com/ethereum/go-ethereum/consensus/dbft.New.func4({0x0, 0xc75aa85fa0766dab?, 0x1304c9266330c603?})
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:326 +0x216
github.com/nspcc-dev/dbft.(*Context[...]).reset(0x1f8e0a0, 0x0, 0x2c)
	github.com/nspcc-dev/dbft@v0.2.0/context.go:201 +0xba
github.com/nspcc-dev/dbft.(*DBFT[...]).initializeConsensus(0x1f92ba0, 0x0, 0xc00012afc8)
	github.com/nspcc-dev/dbft@v0.2.0/dbft.go:88 +0x45
github.com/nspcc-dev/dbft.(*DBFT[...]).Start(0x1f92ba0, 0x0)
	github.com/nspcc-dev/dbft@v0.2.0/dbft.go:74 +0x67
github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).Start(0x0?, {0x1f78150?, 0xc000427808?})
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1133 +0x3eb
created by github.com/ethereum/go-ethereum/eth.(*Ethereum).StartMining in goroutine 1
	github.com/ethereum/go-ethereum/eth/backend.go:524 +0x6c9
```


I'm fixing it. @txhsl, @chenquanyu from your side regarding base fee changes I need you to take a look at https://github.com/ethereum/go-ethereum/pull/28933 and https://github.com/ethereum/go-ethereum/pull/29005 and check whether we need to adjust fees related code. From my PoW we can go without any additional changes, but may be I'm wrong.

txhsl commented:
> @txhsl, @chenquanyu from your side regarding base fee changes I need you to take a look at https://github.com/ethereum/go-ethereum/pull/28933 and https://github.com/ethereum/go-ethereum/pull/29005 and check whether we need to adjust fees related code.

Yes, I think we can go without any additional changes. I'm glad to see that a minimum fee setting was finally applied to the blob pool, it made the codes reasonable.

But before we enable the blob transactions when release v0.2.2, we need to apply our policy limit to the pool properly, ref #266.

AnnaShaleva commented:
> panic: failed to retrieve next block validators: failed to perform 'getCurrentConsensus' call: execution reverted

Regarding this problem:
1. Mainnet/testnet synchronisation fails with the same error on the top of the current version of PR.
2. I tried to locate the source of this error, and it seems to me that it goes from the Governance contract code itself. Given the fact that privnet is running OK on both current master (eef0b9643f65ca5b508da7b5c40051c29d8d7785) and on old branch with v1.13.15 changes on the top of it (e828bfcac3ab73a138b1939de5ea8f047824ff96) I suspect that something has been changed in both Geth v1.13.11+ and compiled Governance version which leads to the execution revert.
3. I located the commit starting from which privnet fails with this error. It's ee1cfc8, in this commit Governance contract (and a couple of other contracts) were updated to fetch recent changes.
4. Now I'm trying to locate commit from Geth v1.13.11+ that leads to this error and connected with changes in Governance introduced by ee1cfc8. I suppose that it should be some change in the VM opcode processing, may be yet another bug fix, or something related to Shanghai opcodes processing, because ee1cfc8 includes a compiler target switch (to Shanghai).

txhsl commented:
> panic: failed to retrieve next block validators: failed to perform 'getCurrentConsensus' call: execution reverted

It is hard to figure out the reason, since the only log we can see is `execution reverted`. I'll check this contract calling from Cancun version as well.

AnnaShaleva commented:
> I'll check this contract calling from Shanghai version as well.

Just to clarify: I don’t think it’s the contract’s fault. It looks like something has been changed in the VM processors.

txhsl commented:
Just some results, I tried several test cases about `eth_Call` based on both `bane-main` and this branch. I sent `getCurrentConsensus()` request to different addresses, so the method hash is `0x9f9d7f81`.

||Address without any code|Contract with the method|Proxy contract without the method|Normal contract without the method|
|--|--|--|--|--|
|`bane-main`|`0x`|`{correct data}`|`execution reverted`|`execution reverted`|
|`upd-v1.13.15`|`0x`|`execution reverted`|`execution reverted`|`invalid opcode: PUSH0`|

txhsl commented on params/config.go:
```diff
@@ -970,9 +973,9 @@ func (c *ChainConfig) Rules(num *big.Int, isMerge bool, timestamp uint64) Rules
-		IsCancun:         c.IsCancun(num, timestamp),
-		IsPrague:         c.IsPrague(num, timestamp),
-		IsVerkle:         c.IsVerkle(num, timestamp),
+		IsShanghai:       isMerge && c.IsShanghai(num, timestamp),
```
Delete this `isMerge &&` and things will work.

AnnaShaleva replied:
Oh gosh...You're right, it works.

AnnaShaleva committed:
params: allow post-merge hardforks to be enabled before the merge

Partially revert https://github.com/ethereum/go-ethereum/pull/29023.
This commit intentionally breaks the post-merge hardforks logic to fix
the problem described in
https://github.com/bane-labs/go-ethereum/pull/232#issuecomment-2278339286.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
Let me do the final checks with mainnet/testnet sync.

AnnaShaleva commented:
Mainnet sync passes:
```
2024-08-14T19:27:06.130+0300	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 2, "height": 144802, "view": 0, "my_height": 144802, "my_view": 0}
2024-08-14T19:27:06.130+0300	INFO	dbft@v0.2.0/dbft.go:498	received Commit	{"validator": 2}
2024-08-14T19:27:06.131+0300	INFO	dbft@v0.2.0/check.go:65	approving block	{"height": 144802, "hash": "0xf98bcbefc6f8b9972d904dbe18e0ddba60d5750c166b29420f8c92e469bd070e", "tx_count": 0, "merkle": "0x6c42725a23c7319177fbee1a7b737684ea07a1268e565766efd30983d9d62082", "prev": "0x86cdb96a0f2139db487f326f4c9cb831b0b65a984ac18d8909e32921ba04bc91"}
INFO [08-14|19:27:06.141] Chain reorg detected                     number=144,800 hash=2363f2..b0269d drop=1 dropfrom=5fea5a..bdd62e add=2 addfrom=9b648e..79888f
INFO [08-14|19:27:06.143] Successfully wrote new block with state  number=144,802 hash=9b648e..79888f
INFO [08-14|19:27:06.144] New block in the chain                   "dbft index"=144,802 "chain index"=144,802 hash=0x9b648e17fb85f30d3883fdabe6ed39568182b2f44e4c5377cad2406cd879888f "parent hash"=0x86cdb96a0f2139db487f326f4c9cb831b0b65a984ac18d8909e32921ba04bc91 primary=0 coinbase=0x1212000000000000000000000000000000000003 "mix digest"=0xeb59c093e3a02bfa4e0d4677d4769022cd9399bbc8b93ad1e892acd6a08aa533
INFO [08-14|19:27:06.144] Fetching latest sealing proposal         "desired number"=144,803
INFO [08-14|19:27:06.145] Commit new sealing work                  number=144,803 sealhash=416eab..b0ee09 "parent hash"=9b648e..79888f etherbase=0x1212000000000000000000000000000000000003 txs=0   gas=0 fees=0 elapsed=1.316ms
```
Testnet sync passes, although testnet sync is significantly slower than it should be due to #296, will fix it in another PR.
```
2024-08-14T20:04:37.689+0300	DEBUG	dbft@v0.2.0/check.go:56	not enough to commit	{"count": 3}
2024-08-14T20:04:37.704+0300	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 0, "height": 262743, "view": 0, "my_height": 262743, "my_view": 0}
2024-08-14T20:04:37.705+0300	INFO	dbft@v0.2.0/dbft.go:498	received Commit	{"validator": 0}
2024-08-14T20:04:37.705+0300	DEBUG	dbft@v0.2.0/check.go:56	not enough to commit	{"count": 4}
2024-08-14T20:04:37.709+0300	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 262743, "view": 0, "my_height": 262743, "my_view": 0}
2024-08-14T20:04:37.710+0300	INFO	dbft@v0.2.0/dbft.go:498	received Commit	{"validator": 3}
2024-08-14T20:04:37.710+0300	INFO	dbft@v0.2.0/check.go:65	approving block	{"height": 262743, "hash": "0xc57abd59633ab12a21ba5cbe948d43c461b2bc014687588b238cb47e87448ae8", "tx_count": 0, "merkle": "0xfcf7c365f64b3b366b57f37a40b05932cf23bf8d129accefb97eacea3c7f7424", "prev": "0xc7f1bda8c45e4b284075e3d487cf9168cbbae905dc5a6b2b08da1fce627e766a"}
INFO [08-14|20:04:37.720] Chain reorg detected                     number=262,741 hash=cd2b8e..6dfdc2 drop=1 dropfrom=23a935..770ad2 add=2 addfrom=d207b3..da9204
INFO [08-14|20:04:37.722] Successfully wrote new block with state  number=262,743 hash=d207b3..da9204
INFO [08-14|20:04:37.722] New block in the chain                   "dbft index"=262,743 "chain index"=262,743 hash=0xd207b35ffecc3d369433d342219d7e895ca43dcce8570871c8160824bada9204 "parent hash"=0xc7f1bda8c45e4b284075e3d487cf9168cbbae905dc5a6b2b08da1fce627e766a primary=5 coinbase=0x1212000000000000000000000000000000000003 "mix digest"=0x072bc064323344cba6d63cad4ca88afbea585fc612919e3e351f457ea3704f76
INFO [08-14|20:04:37.722] Fetching latest sealing proposal         "desired number"=262,744
INFO [08-14|20:04:37.724] Commit new sealing work                  number=262,744 sealhash=adf000..a9943a "parent hash"=d207b3..da9204 etherbase=0x1212000000000000000000000000000000000003 txs=0    gas=0 fees=0 elapsed=1.808ms
INFO [08-14|20:04:38.724] Sealing proposal updated                 number=262,744 sealhash=adf000..a9943a "parent hash"=d207b3..da9204 txs=0
2024-08-14T20:04:38.725+0300	INFO	dbft@v0.2.0/dbft.go:107	initializing dbft	{"height": 262744, "view": 0, "index": -1, "role": "WatchOnly"}
2024-08-14T20:04:38.725+0300	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 1, "height": 262743, "view": 0, "my_height": 262744, "my_view": 0}
2024-08-14T20:04:38.725+0300	DEBUG	dbft@v0.2.0/dbft.go:223	ignoring old height	{"height": 262743}
INFO [08-14|20:04:47.567] Looking for peers                        peercount=1 tried=215 static=0
```
Ready for the merge.

AnnaShaleva commented:
@txhsl, @roman-khimov, please review&merge.

roman-khimov reviewed (approved):
Can't say much about upstream changes, but adaptation is rather clear and if tests are fine then it's OK.

txhsl commented:
Will merge this after some necessary tests from our side.
